contracts/modules/UniswapManager.sol

```diff
@@ -0,0 +1,159 @@
+
+    // The Uniswap factory contract
+    UniswapFactory uniswap;
+    // The Guardian storage 
+    GuardianStorage public guardianStorage;
+
+    /**
+     * @dev Throws if the wallet is locked.
+     */
+    modifier onlyWhenUnlocked(BaseWallet _wallet) {
+        require(!guardianStorage.isLocked(_wallet), "TT: wallet must be unlocked");
+        _;
+    }
+
+    constructor(
+        ModuleRegistry _registry, 
+        GuardianStorage _guardianStorage, 
+        address _uniswap
+    ) 
+        BaseModule(_registry, NAME) 
+        public 
+    {
+        guardianStorage = GuardianStorage(_guardianStorage);
+        uniswap = UniswapFactory(_uniswap);
+    }
+ 
+    /**
+     * @dev Adds liquidity to a Uniswap ETH-ERC20 pair.
+     * @param _wallet The target wallet
+     * @param _poolToken The address of the ERC20 token of the pair.
+     * @param _ethAmount The amount of ETH available.
+     * @param _tokenAmount The amount of ERC20 token available.
+     * @return the number of liquidity shares minted. 
+     */
+    function addLiquidityToUniswap(
+        BaseWallet _wallet, 
+        address _poolToken, 
+        uint256 _ethAmount, 
+        uint256 _tokenAmount
+    )
+        external 
+        onlyOwner(_wallet)
+        onlyWhenUnlocked(_wallet)
+    {
+        address pool = uniswap.getExchange(_poolToken);
+        require(pool != address(0), "UM: The target token is not traded on Uniswap");
+
+        uint256 ethPoolSize = address(pool).balance;
+        uint256 tokenPoolSize = ERC20(_poolToken).balanceOf(pool);
+
+        uint256 tokenInEth = getInputToOutputPrice(_tokenAmount, tokenPoolSize, ethPoolSize);
+        if(_ethAmount >= tokenInEth) {
+            // swap some eth for tokens
+            (uint256 ethSwap, uint256 ethPool, uint256 tokenPool) = computePooledValue(ethPoolSize, tokenPoolSize, _ethAmount, _tokenAmount, tokenInEth);
+            _wallet.invoke(pool, ethSwap, abi.encodeWithSignature("ethToTokenSwapInput(uint256,uint256)", 1, block.timestamp));
+            _wallet.invoke(_poolToken, 0, abi.encodeWithSignature("approve(address,uint256)", pool, tokenPool));
+            // add liquidity
+            _wallet.invoke(pool, ethPool, abi.encodeWithSignature("addLiquidity(uint256,uint256,uint256)",1, tokenPool, block.timestamp + 1));
+        }
+        else {
+            // swap some tokens for eth
+            (uint256 tokenSwap, uint256 tokenPool, uint256 ethPool) = computePooledValue(tokenPoolSize, ethPoolSize, _tokenAmount, _ethAmount, 0);
+            _wallet.invoke(_poolToken, 0, abi.encodeWithSignature("approve(address,uint256)", pool, tokenSwap + tokenPool));
+            _wallet.invoke(pool, 0, abi.encodeWithSignature("tokenToEthSwapInput(uint256,uint256,uint256)", tokenSwap, 1, block.timestamp));
+            // add liquidity
+            _wallet.invoke(pool, ethPool - 1, abi.encodeWithSignature("addLiquidity(uint256,uint256,uint256)",1, tokenPool, block.timestamp + 1));
```

Comment:
why `ethPool - 1` ?

Reply:
We are trying to satisfy the condition at L57 of uniswap_exchange.vy which translates to:
```tokenPool >= ethPool * token_reserve / eth_reserve + 1``` 
The `+1` in the expression above comes from the integer division.

The `ethPool - 1` and `ethPool + 1` where naive attempts to take this `+1` into account. 

To be more consistent I have now updated the code to replace `tokenPool` by `tokenPool + 1` when the quantities have been computed.